graphbolt/include/csr_sampling_graph.h

```diff
@@ -0,0 +1,131 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ * @file graphbolt/include/csr_sampling_graph.h
+ * @brief Header file of csr sampling graph.
+ */
+
+#include <torch/custom_class.h>
+#include <torch/torch.h>
+
+#include <string>
+#include <vector>
+
+namespace graphbolt {
+namespace sampling {
+
+using StringList = std::vector<std::string>;
+
+/**
+ * @brief Heterogeneous information in the Graph.
+ */
+struct HeteroInfo {
+  HeteroInfo(
+      const StringList& ntypes, const StringList& etypes,
+      torch::Tensor& node_type_offset_, torch::Tensor& type_per_edge_)
+      : node_types(ntypes),
+        edge_types(etypes),
+        node_type_offset(node_type_offset_),
+        type_per_edge(type_per_edge_) {}
+  /** @brief List of node types in the graph. */
+  StringList node_types;
+  /** @brief List of edge types in the graph. */
+  StringList edge_types;
+  /** @brief Offset array of node type. */
+  torch::Tensor node_type_offset;
+  /** @brief Type of each edge. */
+  torch::Tensor type_per_edge;
+};
+
+/**
+ * @brief A sampling oriented csr format graph.
+ */
+class CSRSamplingGraph : public torch::CustomClassHolder {
+ public:
+  /**
+   * @brief Constructor for CSR with data.
+   * @param num_rows The number of rows in the dense shape.
+   * @param num_cols The number of columns in the dense shape.
+   * @param indptr The CSR format index pointer array.
+   * @param indices The CSR format index array.
+   * @param hetero_info Heterogeneous graph information, if present.
+   */
+  CSRSamplingGraph(
+      int64_t num_rows, int64_t num_cols, torch::Tensor& indptr,
+      torch::Tensor& indices, const std::shared_ptr<HeteroInfo>& hetero_info);
+
+  /**
+   * @brief Create a CSR graph from tensors of CSR format.
+   * @param indptr Index pointer array of the CSR.
+   * @param indices Indices array of the CSR.
+   * @param shape Shape of the sparse matrix.
+   *
+   * @return CSRSamplingGraph
+   */
+  static c10::intrusive_ptr<CSRSamplingGraph> FromCSR(
+      const std::vector<int64_t>& shape, torch::Tensor indptr,
+      torch::Tensor indices);
+
+  /**
+   * @brief Set the heterogeneous information for the graph.
+   *
+   * @param ntypes A list of node types.
+   * @param etypes A list of edge types.
+   * @param node_type_offset_ A tensor representing the offset of node types.
+   * @param type_per_edge_ A tensor representing the type of each edge.
+   */
+  void SetHeteroInfo(
```

Comment:
Make sense, it is originally used for test, let me remove it.

Reply:
I see. For API simplicity, we may have two interfaces for SamplingGraph construction in the Python side: `from_homo_csc()` and `from_hetero_csc()`. The first one is called without heterogeneous information while the latter one also takes the arguments in the current `SetHeteroInfo` as input. Anyway, I don't like `SetHeteroInfo` because it implies that the heterogeneous of a graph is mutable.

Reply:
Make sense, chang for fix.